fedimint-core/src/query.rs

```diff
@@ -163,100 +168,67 @@ impl<R: Debug + Eq + Clone> QueryStrategy<R, Vec<R>> for UnionResponsesSingle<R>
-    required: usize,
+    /// Previously received responses/errors
+    responses: BTreeMap<PeerId, R>,
+    errors: BTreeMap<PeerId, MemberError>,
+    responded_peers: BTreeSet<PeerId>,
+    threshold: usize,
+    max_evil: usize,
 }
 
 impl<R> CurrentConsensus<R> {
-    pub fn new(required: usize) -> Self {
+    pub fn new(total_peers: usize) -> Self {
+        let max_evil = (total_peers - 1) / 3;
+        let threshold = total_peers - max_evil;
+
         Self {
-            existing_results: vec![],
+            responses: BTreeMap::new(),
             errors: BTreeMap::new(),
-            required,
+            responded_peers: BTreeSet::new(),
+            threshold,
+            max_evil,
         }
     }
 }
 
 impl<R: Eq + Clone + Debug> QueryStrategy<R> for CurrentConsensus<R> {
     fn process(&mut self, peer: PeerId, result: api::MemberResult<R>) -> QueryStep<R> {
         match result {
-            Ok(result) => {
-                if let Some((prev_result, peers)) = self
-                    .existing_results
-                    .iter_mut()
-                    .find(|(prev_result, _)| prev_result == &result)
-                {
-                    if peers.contains(&peer) {
-                        debug!(prev = ?prev_result, new = ?result, peer = %peer, "Ignoring duplicate response from peer");
-                    } else {
-                        peers.insert(peer);
-                    }
-                } else {
-                    self.existing_results.push((result, HashSet::from([peer])));
-                }
+            Ok(response) => {
+                self.responses.insert(peer, response);
+                self.responded_peers.insert(peer);
             }
             Err(error) => {
                 self.errors.insert(peer, error);
+
+                if self.errors.len() > self.max_evil {
+                    return QueryStep::Failure {
+                        general: None,
+                        members: mem::take(&mut self.errors),
+                    };
+                }
             }
         }
 
-        for (result, peers) in &self.existing_results {
-            if peers.len() >= self.required {
-                return QueryStep::Success(result.clone());
+        if let Some(response) = self
+            .responses
+            .values()
+            .max_by_key(|response| self.responses.values().filter(|r| r == response).count())
+        {
+            let count = self.responses.values().filter(|r| r == &response).count();
+
+            if count >= self.threshold {
+                return QueryStep::Success(response.clone());
             }
         }
 
-        if self.errors.len() >= self.required {
-            return QueryStep::Failure {
-                general: None,
-                members: mem::take(&mut self.errors),
-            };
+        if self.responded_peers.len() >= self.threshold {
+            QueryStep::RetryMembers(mem::take(&mut self.responded_peers))
```

Comment:
I'm not super familiar with this part of the code. What will `RetryMembers` do? Is there some sort of backoff?

Reply:
If we return RetryMembers the query will call the endpoint again and call process on the result. this allows peers to update their value in order for us to achieve a quorum.

Reply:
Is there a backoff and timeout?

Reply:
Yes there is a initial delay of 10ms that doubles on every retry to the same peer.